### DISCUSSION TITLE
Datto : Setting up API Data Source

### DISCUSSION THREAD

**Discussion by PredictConnor** [grafana/grafana-infinity-datasource#124] [Specific APIs]
I'm going to start out by apologising as I'm completely out of my depth here but refuse to give up.

I have various systems im trying to pull together and grafana with Infinity seems to be able to provide exactly what I'm after however my lack of ability is preventing me moving forward.

An example I use Datto 
They provide me the following.

API URL
API Key
API Secret
All of the variables required to pull data.

I seem completely unable to even authenticate to the API, I can get it working correctly in postman just no in Infinity.

If anyone can point me in the right direction I'd be so greatful!

Do I I need basic auth? Headers defined? I'm so clueless.
Thanks to anyone who has the patience to help me :)

**Comment by yesoreyeram** [comment]
Hey, From the  datto doc I understand it require OAUTH workflow. The plugin doesn't support outh workflow to fetch a token. Plugin only supports username/password authentication or header based authentication. 

I don't have any datto account to test. So I might be completley wrong.

**Reply by PredictConnor** [reply]
Thanks!

Gotcha! That seems to make sense.
I suppose, I'll try and find a plugin that allows OAuth then :)
If you know of any let me know!

**Comment by csthomas2** [comment]
I got this to work by first setting up authentication in postman. You can use Postman's console to check the header it sends when you send a request to Datto's API.

I went to my Infinity data source and added a header titled "Authentication" with the value set to "Bearer MYAPIKEY". It'll work until the key expires.

What I can't seem to do, is create a time series out of the information Datto sends back.

**Reply by PredictConnor** [reply]
That sounds amazing! Honestly this small idea has turned into a whole world of developer that I'm not in haha!
I'll have a play with this later on today and let you know how I get on!

I did use postman to test my auth recently and it presented me a Key + Secret Key display box but I'll give it ago again and spend some more time

**Reply by PredictConnor** [reply]
Hi Csthomas,

I've managed to get the  API working correctly within Postman, however can't for the life of me getting it working in Grafana.
Infinity Plugin.
url blank

Header - Authorization
Value - Bearer eyJhbGciOiJ (Rest of my Generated Bearer Key)

API Request - https://pinotage-api.centrastage.net/api/v2/account/alerts/open/

All this works in postman just not Grafana most strange, any clues

**Reply by csthomas2** [reply]
Hi PredictConnor,

I was only able to get this far using this plugin because of 2 major limitations. I can't format the data from the API into a time series so I can't display it in anything but a table like Excel. The second is that since the OAUTH2 token expires, it won't renew unless I update it manually. I decided to use Python to gather then expose my custom metrics with Python's Prometheus library. 

To answer your question though, in the Edit Panel I have table view switched on at the top (or the visualization set to table), the URL set as you do without the trailing slash https://pinotage-api.centrastage.net/api/v2/account/alerts/open and beside the datasource (set to infinity) the query options min interval set to 1s)

In the data source configuration, I left Base URL blank, and only filled in the extra header for authorization the same way as you did.

**Reply by PredictConnor** [reply]
Weird that.. Getting no errors, but no data is being displayed.
If I run a query inspector it doesn't load.

I have seen the issue with the OAUTH expiration which is annoying but was going to tackle the first issue of viewing the data.
I do the same btw, leave the URL blank in the infinity plugin. 

So annoying as I've managed to get my other systems integrated now :(!

**Reply by PredictConnor** [reply]
Evening,

I finally managed to get the API working within Grafana. The JSON format the API pulls is very bizarre and as you described is hard to format. I've managed to bodge a solution that shows my alerts, Offline Devices etc as a counter however anything past that seems impossible. The OAuth time out weirdly hasn't kicked in for me and it's been running for 6 days.. 

The formatting though it pulls due to the format seems to table nearly everything together.

Quite underwhelming to be honest!

**Reply by solly4by7** [reply]
Hi I face a similar issue with REST API's and OAuth2. Can you let me know as to how you had manged to pull data to Grafana through REST API's.

**Reply by mit-corp** [reply]
Hello,

I'm in the same situation. 

Postman allows us to use Oauth 2.0 (as described in the DattoRMM documentation).
Except that Postman offers the possibility of transferring authentication to the header:  "Send as Basic Auth header"
 (= Authorization: Bearer XXXXXXXXXXXXX)

Maybe in the next-version you could be implement that ? :)
